### ISSUE TITLE
Support `Upgrade` requests (Socket.IO)

### ISSUE THREAD

**Issue by kettanaito** [mswjs/interceptors#682]
We don't seem to passthrough Socket.IO `Upgrade` HTTP request correctly. It gets stuck forever.

**Comment by kettanaito** [comment]
Here's the `net.Socket` emit/write log comparison between the working (no interceptor) and broken (ClientRequest interceptor applied) versions:

## No interceptor

```
WRITE! GET /socket.io/?EIO=4&transport=websocket HTTP/1.1
Sec-WebSocket-Version: 13
Sec-WebSocket-Key: EJ2fmeWdlAdTAA6wxdBcIg==
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
Host: localhost:51678


EMIT! resume

EMIT! lookup

EMIT! resume
EMIT! connect
EMIT! ready

WRITE! HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: 9y6umGlFvZE0sLuAn73qjWjwkiw=


EMIT! resume
WRITE! �k
WRITE! 0{"sid":"Kz0Rh53n7Ogm3NUyAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000}

EMIT! data
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: 9y6umGlFvZE0sLuAn73qjWjwkiw=

�k0{"sid":"Kz0Rh53n7Ogm3NUyAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000} 


EMIT! agentRemove
EMIT! readable
EMIT! data
�k0{"sid":"Kz0Rh53n7Ogm3NUyAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000} 


WRITE! �����
WRITE! ��
EMIT! resume
EMIT! data
������� 


WRITE! � 
WRITE! 40{"sid":"f3xP3tuJ8TKnj00bAAAB"}
EMIT! data
� 40{"sid":"f3xP3tuJ8TKnj00bAAAB"} 


---- (connection open, test done)

WRITE! �
WRITE! 41
WRITE! �
WRITE! 

�41� 




WRITE! ���Ī�
WRITE! 


���Ī� 




EMIT! data
EMIT! prefinish
EMIT! finish
EMIT! data
EMIT! prefinish
EMIT! finish
EMIT! readable
EMIT! end
EMIT! readable
EMIT! end
EMIT! close
EMIT! close
```

## Interceptor applied

```
EMIT! resume

WRITE! GET /socket.io/?EIO=4&transport=websocket HTTP/1.1
Sec-WebSocket-Version: 13
Sec-WebSocket-Key: u0avYGhp4h9fN8N4E8QsDw==
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
Host: localhost:51678


EMIT! resume
EMIT! resume

EMIT! lookup
EMIT! lookup
EMIT! connect
EMIT! connect
EMIT! ready
EMIT! ready
EMIT! resume
WRITE! HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: 08kvjVAwS1LE6zG3pNIG5uFpeys=


EMIT! resume
WRITE! �k
WRITE! 0{"sid":"eUD8fqOV9Sf6nzTMAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000}

EMIT! data
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: 08kvjVAwS1LE6zG3pNIG5uFpeys=

�k0{"sid":"eUD8fqOV9Sf6nzTMAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000} 


EMIT! data
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: 08kvjVAwS1LE6zG3pNIG5uFpeys=

�k0{"sid":"eUD8fqOV9Sf6nzTMAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000} 


EMIT! agentRemove
EMIT! readable
EMIT! data
�k0{"sid":"eUD8fqOV9Sf6nzTMAAAA","upgrades":[],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000} 


EMIT! resume

WRITE! �
WRITE! 

EMIT! data
EMIT! data

� 


� 


```

There are clearly differences, mostly:

- Interceptor does not receive `40`/`41` payloads from the server, which is Socket.IO's way of acknowledging the connection.
- Ignore the prefinish/finish/end/close sequence missing at the end of the interceptor scenario. It's missing because the connection is never open, so it won't close either (test times out).

**Comment by kettanaito** [comment]
How we handle request passthrough:

https://github.com/mswjs/interceptors/blob/0ce43e8f6b97f1fe6880c6ea2789fb50ffc9e78d/src/interceptors/ClientRequest/MockHttpSocket.ts#L153-L257

**Comment by kettanaito** [comment]
## Released: v0.37.3 🎉

This has been released in v0.37.3!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.37.3)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.37.3)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
